### PR TITLE
Add the example of super_resolution

### LINKED ISSUE #2878: Addition of more tutorial notebooks on topics like Siamese Networks and Super Resolution techniques
## 🚀 Feature
The number of examples in ignite repository is significantly less as compared to the [pytorch/examples](https://github.com/pytorch/examples) repository. It would be great to have a few more Google Colab tutorials on topics like Siamese Networks and Super resolution in images.
The above examples are already a part of [pytorch/examples](https://github.com/pytorch/examples) repository and hence will be helpful to refer and implement.

### PR DESCRIPTION
**Pull request by guptaaryan16**
Fixes #2878 

Description: I have added the example for the super resolution in the `PyTorch/examples`

Check list:

- [x] New tests are added (if a new feature is added)
- [ ] New doc strings: description and/or example code are in RST format
- [ ] Documentation is updated (if required)

### CODE REVIEW

**Comment by guptaaryan16** [comment]
Hey @vfdev-5 , thanks for assigning the issue and since I had my mid-terms I could not work on this earlier so I apologise for the delay

File: examples/super_resolution/README.md [hunk 1]
```diff
@@ -0,0 +1,37 @@
+# Superresolution using an efficient sub-pixel convolutional neural network
```
Comment (vfdev-5) [inline examples/super_resolution/README.md:1]:
```suggestion
# Super-Resolution using an efficient sub-pixel convolutional neural network
```

File: examples/super_resolution/README.md [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+  --testBatchSize       testing batch size
+  --nEpochs             number of epochs to train for
+  --lr                  Learning Rate. Default=0.01
+  --cuda                use cuda
+  --mps                 enable GPU on macOS
+  --threads             number of threads for data loader to use Default=4
+  --seed                random seed to use. Default=123
+```
+
+This example trains a super-resolution network on the [BSD300 dataset](https://www2.eecs.berkeley.edu/Research/Projects/CS/vision/bsds/), using crops from the 200 training images, and evaluating on crops of the 100 test images. A snapshot of the model after every epoch with filename model*epoch*<epoch_number>.pth
```
Comment (vfdev-5) [inline examples/super_resolution/README.md:27]: `model*epoch*<epoch_number>.pth` is this info is correct ?

File: examples/super_resolution/data.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
+import tarfile
+from os import makedirs, remove
+from os.path import basename, exists, join
+
+from dataset import DatasetFromFolder
+from six.moves import urllib
```
Comment (vfdev-5) [inline examples/super_resolution/data.py:6]: Why do you use six here ?

File: examples/super_resolution/main.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
+from __future__ import print_function
```
Comment (vfdev-5) [inline examples/super_resolution/main.py:1]: This is useless for python 3.8+

**Review by vfdev-5 — COMMENTED** [review]
Thanks for the PR @guptaaryan16 
I left few comments.
Can you share execution logs, how much time it takes to run one epoch ?
Were you be able to train a simple SR model and share inference example ?

File: examples/super_resolution/dataset.py [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    def __init__(self, image_dir, input_transform=None, target_transform=None):
+        super(DatasetFromFolder, self).__init__()
+        self.image_filenames = [join(image_dir, x) for x in listdir(image_dir) if is_image_file(x)]
+
+        self.input_transform = input_transform
+        self.target_transform = target_transform
+
+    def __getitem__(self, index):
+        input = load_img(self.image_filenames[index])
+        target = input.copy()
```
Comment (vfdev-5) [inline examples/super_resolution/dataset.py:28]: Seems like in this application we are using input image as target, so we can replace BSD data by another built-in torchvision dataset instead of manually code how to download/extract the data

**Comment by guptaaryan16** [comment]
I think this colab notebook will be able to cover this example
https://colab.research.google.com/drive/1mg6TzaUOgy8CJ7TwGVOKHNZJ1UhE6u9E?usp=sharing

File: examples/super_resolution/README.md (hunk 2 above)
Comment (guptaaryan16) [reply examples/super_resolution/README.md]: Yes but the correct storage path is model_epoch_<epoch_number>.pth

**Comment by guptaaryan16** [comment]
A lot of things covered in this example were directly present in `PyTorch/examples` repo of the same example so it may contain some similar errors

File: examples/super_resolution/dataset.py (hunk 5 above)
Comment (guptaaryan16) [reply examples/super_resolution/dataset.py:28]: Although it is working well as shown in Colab notebook

**Comment by vfdev-5** [comment]
@guptaaryan16 code in pytorch/examples seems a bit outdated and rather bulky. For ignite, I think it would worth to provide up to date code and with interesting outcome.

First question about the results from the colab, how super-resolved image differs from an image upsampled with bicubic mode ? Maybe, it worth to include PSNR for bicubic upsampled image for reference when doing validation stage.
It would be also funny to super-resolve for test images those from cifar10.

**Comment by guptaaryan16** [comment]
Hey @vfdev-5 I have made all suggested changes including adding the torchvision dataset. Please review

File: examples/super_resolution/main.py [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+
+print("===> Loading datasets")
+
+
+class SRDataset(torch.utils.data.Dataset):
+    def __init__(self, dataset, scale_factor):
+        self.dataset = dataset
+        self.transform = transforms.Resize(
+            (len(dataset[0][0][0]) * scale_factor, len(dataset[0][0][0][0]) * scale_factor)
+        )
```
Comment (vfdev-5) [inline examples/super_resolution/main.py:52]: I think you can just use torchvision functional API instead of this:
```
from torchvision.transforms.functional import resize
```

File: examples/super_resolution/main.py [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+
+@trainer.on(Events.EPOCH_COMPLETED(every=validate_every))
+def run_validation():
+    evaluator.run(testing_data_loader)
+
+
+@trainer.on(Events.EPOCH_COMPLETED(every=validate_every))
+def log_validation():
+    metrics = evaluator.state.metrics
+    print(f"Epoch: {trainer.state.epoch}, Avg. PSNR: {metrics['psnr']} dB")
```
Comment (vfdev-5) [inline examples/super_resolution/main.py:128]:
```suggestion
@trainer.on(Events.EPOCH_COMPLETED(every=validate_every))
def run_validation():
    evaluator.run(testing_data_loader)
    metrics = evaluator.state.metrics
    print(f"Epoch: {trainer.state.epoch}, Avg. PSNR: {metrics['psnr']} dB")
```

File: examples/super_resolution/main.py [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+
+from ignite.engine import Engine, Events
+from ignite.metrics import PSNR
+
+# Training settings
+parser = argparse.ArgumentParser(description="PyTorch Super Res Example")
+parser.add_argument("--upscale_factor", type=int, required=True, help="super resolution upscale factor")
+parser.add_argument("--batchSize", type=int, default=64, help="training batch size")
+parser.add_argument("--testBatchSize", type=int, default=10, help="testing batch size")
+parser.add_argument("--nEpochs", type=int, default=2, help="number of epochs to train for")
```
Comment (vfdev-5) [inline examples/super_resolution/main.py:19]: Let's use snake case here : `--batch_size` and `--test_batch_size` etc

File: examples/super_resolution/main.py (hunk 6 above)
Comment (vfdev-5) [reply examples/super_resolution/main.py:52]: So, looks like you are upsampling input image with bilinear mode. Model will learn how to perform bilinear upsampling which is not super-resolution...

I think my previous point about CIFAR10 was about using it for evaluation only and not training. We should train on large images rescaled to small ones (as dataset you can pick one among available, scenes/classes similar to CIFAR10 and downloadable on torchvision: CalTech101, VOC, etc, https://pytorch.org/vision/stable/datasets.html) 

Then we can visually check generalization on CIFAR10.

**Review by vfdev-5 — COMMENTED** [review]
Thanks for the update @guptaaryan16 

I left few comments in the code. I think we can keep this model and the training script but change the training dataset and use additionally CIFAR10 for inference only.

File: examples/super_resolution/main.py (hunk 6 above)
Comment (guptaaryan16) [reply examples/super_resolution/main.py]: Hey @vfdev-5 I made changes to get the images rescaled to smaller ones during training as present in the `PyTorch/examples` and suggested by you.

File: examples/super_resolution/main.py [hunk 9]
```diff
@@ -45,32 +45,71 @@
...
+def calculate_valid_crop_size(crop_size, upscale_factor):
+    return crop_size - (crop_size % upscale_factor)
+
+
+def input_transform(crop_size, upscale_factor):
+    return Compose(
+        [
+            CenterCrop(crop_size),
+            Resize(crop_size // upscale_factor),
+            ToTensor(),
```
Comment (vfdev-5) [inline examples/super_resolution/main.py:78]: @guptaaryan16 why do you crop input images instead of simply resizing them with functional API ?

File: examples/super_resolution/README.md [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+  --upscale_factor      super resolution upscale factor
+  --batchSize           training batch size
+  --testBatchSize       testing batch size
+  --nEpochs             number of epochs to train for
+  --lr                  Learning Rate. Default=0.01
+  --cuda                use cuda
+  --mps                 enable GPU on macOS
+  --threads             number of threads for data loader to use Default=4
+  --seed                random seed to use. Default=123
+```
```
Comment (vfdev-5) [inline examples/super_resolution/README.md:25]: Still camel case here instead of snake case

File: examples/super_resolution/main.py (hunk 9 above)
Comment (guptaaryan16) [reply examples/super_resolution/main.py:78]: @vfdev-5 the images were of variable size between 200-300 pixels in Caltech 101 so it was giving stack error in the training without cropping them

File: examples/super_resolution/super_resolve.py [hunk 11]
```diff
@@ -0,0 +1,42 @@
+from __future__ import print_function
```
[the remark this answers is not in the document]
Comment (guptaaryan16) [reply examples/super_resolution/super_resolve.py:1]: @vfdev-5 Let me make the changes

File: examples/super_resolution/README.md [hunk 12]
```diff
@@ -0,0 +1,37 @@
+# Super-Resolution using an efficient sub-pixel convolutional neural network
+
+ported from [pytorch-examples](https://github.com/pytorch/examples/tree/main/super_resolution)
+
+This example illustrates how to use the efficient sub-pixel convolution layer described in ["Real-Time Single Image and Video Super-Resolution Using an Efficient Sub-Pixel Convolutional Neural Network" - Shi et al.](https://arxiv.org/abs/1609.05158) for increasing spatial resolution within your network for tasks such as superresolution.
+
+```
+usage: main.py [-h] --upscale_factor UPSCALE_FACTOR [--batchSize BATCHSIZE]
+               [--testBatchSize TESTBATCHSIZE] [--nEpochs NEPOCHS] [--lr LR]
+               [--cuda] [--threads THREADS] [--seed SEED]
```
Comment (vfdev-5) [inline examples/super_resolution/README.md:10]: How about here ?

File: examples/super_resolution/super_resolve.py [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+parser.add_argument("--cuda", action="store_true", help="use cuda")
+opt = parser.parse_args()
+
+print(opt)
+img = Image.open(opt.input_image).convert("YCbCr")
+y, cb, cr = img.split()
+
+model = torch.load(opt.model)
+img_to_tensor = ToTensor()
+input = img_to_tensor(y).view(1, -1, y.size[1], y.size[0])
```
Comment (vfdev-5) [inline examples/super_resolution/super_resolve.py:22]: Same here, use functional API instead of transform

**Comment by guptaaryan16** [comment]
Hey @vfdev-5 please review

File: examples/super_resolution/main.py [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
+
+    def __getitem__(self, index):
+        image, _ = self.dataset[index]
+        img = image.convert("YCbCr")
+        hr_image, _, _ = img.split()
+        hr_image = center_crop(hr_image, self.crop_size)
+        lr_image = hr_image.copy()
+        if self.scale_factor != 1:
+            dim = self.crop_size // self.scale_factor
+            lr_image = resize(lr_image, [dim, dim])
```
Comment (vfdev-5) [inline examples/super_resolution/main.py:61]:
```suggestion
            size = self.crop_size // self.scale_factor
            lr_image = resize(lr_image, [size, size])
```

File: examples/super_resolution/main.py [hunk 15]
```diff
@@ -0,0 +1,149 @@
...
+
+trainset = torchvision.datasets.Caltech101(root="./data", download=True)
+testset = torchvision.datasets.Caltech101(root="./data", download=False)
+
+trainset_sr = SRDataset(trainset, scale_factor=opt.upscale_factor)
+testset_sr = SRDataset(testset, scale_factor=opt.upscale_factor)
+
+training_data_loader = DataLoader(dataset=trainset_sr, num_workers=opt.threads, batch_size=opt.batch_size, shuffle=True)
+testing_data_loader = DataLoader(
+    dataset=testset_sr, num_workers=opt.threads, batch_size=opt.test_batch_size, shuffle=False
```
Comment (vfdev-5) [inline examples/super_resolution/main.py:78]:
```suggestion
    dataset=testset_sr, num_workers=opt.threads, batch_size=opt.test_batch_size
```

File: examples/super_resolution/main.py [hunk 16]
```diff
@@ -0,0 +1,149 @@
...
+
+print("===> Building model")
+model = Net(upscale_factor=opt.upscale_factor).to(device)
+criterion = nn.MSELoss()
+
+optimizer = optim.Adam(model.parameters(), lr=opt.lr)
+
+
+def train_step(engine, batch):
+    input, target = batch[0].to(device), batch[1].to(device)
```
Comment (vfdev-5) [inline examples/super_resolution/main.py:89]:
```suggestion
    model.train()
    input, target = batch[0].to(device), batch[1].to(device)
```

File: examples/super_resolution/super_resolve.py [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+y, cb, cr = img.split()
+
+model = torch.load(opt.model)
+input = to_tensor(y).view(1, -1, y.size[1], y.size[0])
+
+if opt.cuda:
+    model = model.cuda()
+    input = input.cuda()
+
+out = model(input)
```
Comment (vfdev-5) [inline examples/super_resolution/super_resolve.py:27]:
```suggestion
model.eval()
with torch.no_grad():
    out = model(input)
```

**Review by vfdev-5 — COMMENTED** [review]
Thanks for the updates @guptaaryan16 
I left few more comments concerning the code.
Please try to train a model and share training logs and run super_resolve script on a cifar10 image then compare it with the same image upsampled with bicubic interp mode.

**Comment by guptaaryan16** [comment]
Hey @vfdev-5 I have updated the colab link with the necessary training logs and updates along with the cifar-10 example
https://colab.research.google.com/drive/1mg6TzaUOgy8CJ7TwGVOKHNZJ1UhE6u9E?usp=sharing

**Review by vfdev-5 — APPROVED** [review]
LGTM, thanks @guptaaryan16 
I approve this PR. Next step would be to integrate it into the CI and use other ignite tools like progress bar to show sort of best practices
